### PR TITLE
Iterable Dataloader

### PR DESCRIPTION
**Pull request by ibanesh**
Fairseq2 data pipeline is an iterable without index based random access. 
The current simuleval design doesn't allow us to iterate through such iterable dataloaders without fully loading them into memory (which is inefficient). Fairseq2 data pipeline is an iterable which loads the samples from teh data file one at a time or in batches. To accommodate this lazy loading of data and to be memory efficient, refactoring the way we iterate through the dataset. 

Eventually, it would be memory efficient to switch to iterable dataloaders by default and avoid loading the whole file into memory like we do for most dataloaders now. 

An example iterable dataloader is defined in https://github.com/fairinternal/seamless_communication/pull/61 - changes were tested using this PR.

### CODE REVIEW

File: simuleval/agents/pipeline.py [hunk 1]
```diff
@@ -75,7 +75,9 @@ def push(
                 upstream_states=upstream_states + states_list[:index],
             )
         self.module_list[-1].push(
-            segment, states[-1], upstream_states=upstream_states + states_list[:index]
+            segment,
+            states[-1],
+            upstream_states=upstream_states + states_list[: len(self.module_list[:-1])],
```
Comment (annasun28) [inline simuleval/agents/pipeline.py:80]: just for my understanding, when is `len(self.module_list[:-1])` not equal to `index`?

File: simuleval/evaluator/evaluator.py [hunk 2]
```diff
@@ -233,23 +242,29 @@ def is_finished(self, instance) -> bool:
         return instance.finish_prediction
```
Comment (annasun28) [inline simuleval/evaluator/evaluator.py:242]: Do we still need this method after removing `while not self.is_finished(instance):`?

File: simuleval/options.py [hunk 3]
```diff
@@ -100,7 +100,12 @@ def add_evaluator_args(parser: argparse.ArgumentParser):
         default=-1,
         help="The last index for evaluation.",
     )
-    parser.add_argument("--output", type=str, default=None, help="Output directory")
+    parser.add_argument(
+        "--output",
+        type=str,
+        default=None,
+        help="Output directory. Required if using iterable dataloader.",
```
Comment (annasun28) [inline simuleval/options.py:107]: Also just curious: why required if using iterable dataloader but not otherwise?

**Review by annasun28 — APPROVED** [review]

File: simuleval/agents/pipeline.py (hunk 1 above)
Comment (ibanesh) [reply simuleval/agents/pipeline.py:80]: When the len(self.module_list) = 1.
len(self.module_list[:-1]) will be 0 and `index` will not even get assigned a value in this case, resulting in an exception.

File: simuleval/options.py (hunk 3 above)
Comment (ibanesh) [reply simuleval/options.py:107]: For the typical dataloader (other than "iterable" ones), we create a list of instances and iterate through this list in the evaluator's `__call__` method, the results are stored as a filed in the instance object in this list and at the end of iteration we this list from memory to calculate the metrics/scores. So it doesn't matter if an output arg is specified, without an output arg the instances don't get written to a file but we will still be able to calculate the score/metrics using the results in memory.

But for iterable dataloaders where we are lazy loading the data one at a time, only the current instance is stored in memory and at each iteration we write this instance data to a file if `output` is given. At the end of iterating through all the data, we will load the results into memory using the file we wrote to while iterating and then use this in memory instances to calculate the cumulative scores/metric. So without the `output` arg, we won't write to file and subsequently we wont be able to calculate the score/metrics.

File: simuleval/evaluator/evaluator.py [hunk 4]
```diff
@@ -233,23 +242,29 @@ def is_finished(self, instance) -> bool:
...
-                input_segment = instance.send_source(self.source_segment_size)
-                output_segment = system.pushpop(input_segment)
-                instance.receive_prediction(output_segment)
-                if instance.finish_prediction:
-                    # if instance.finish_prediction where set by the reader,
-                    # source_finished_reading will be set as well. If it is
-                    # set by any of the intermediate components, then we didn't
-                    # end yet. We are going to clear the state and continue
-                    # processing the rest of the input.
-                    system.reset()
```
[the remark this answers is not in the document]
Comment (ibanesh) [reply simuleval/evaluator/evaluator.py:248]: Keeping it as is for now, will change this to old logic if we run into issue when testing parity for wait-k.

**Pull request merged**